components/director/pkg/auth/mtls_token_provider.go

```diff
@@ -0,0 +1,168 @@
+/*
+* Copyright 2020 The Compass Authors
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+ */
+
+package auth
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	oauth "github.com/kyma-incubator/compass/components/director/pkg/oauth"
+
+	httpdirector "github.com/kyma-incubator/compass/components/director/pkg/http"
+	"github.com/kyma-incubator/compass/components/director/pkg/log"
+	httputils "github.com/kyma-incubator/compass/components/system-broker/pkg/http"
+	"github.com/pkg/errors"
+)
+
+// CertificateCache missing godoc
+//go:generate mockery --name=CertificateCache --output=automock --outpkg=automock --case=underscore
+type CertificateCache interface {
+	Get() *tls.Certificate
+}
+
+// MtlsClientCreator is a constructor function for http.Clients
+type MtlsClientCreator func(cache CertificateCache, skipSSLValidation bool, timeout time.Duration) *http.Client
+
+// DefaultMtlsClientCreator is the default http client creator
+func DefaultMtlsClientCreator(cc CertificateCache, skipSSLValidation bool, timeout time.Duration) *http.Client {
+	httpTransport := httpdirector.NewCorrelationIDTransport(&http.Transport{
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: skipSSLValidation,
+			GetClientCertificate: func(_ *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+				return cc.Get(), nil
+			},
+		},
+	})
+
+	return &http.Client{
+		Transport: httpTransport,
+		Timeout:   timeout,
+	}
+}
+
+// mtlsTokenAuthorizationProvider presents a AuthorizationProvider implementation which crafts OAuth Bearer token values for the Authorization header using mtls http client
+type mtlsTokenAuthorizationProvider struct {
+	httpClient *http.Client
+}
+
+// NewMtlsTokenAuthorizationProvider constructs an TokenAuthorizationProvider
+func NewMtlsTokenAuthorizationProvider(oauthCfg oauth.Config, cache CertificateCache, creator MtlsClientCreator) *mtlsTokenAuthorizationProvider {
+	return &mtlsTokenAuthorizationProvider{
+		httpClient: creator(cache, oauthCfg.SkipSSLValidation, oauthCfg.TokenRequestTimeout),
```

Comment:
is there any reason to pass a creator here and not just create the client before creating the tokenprovider and pass a client interface to the tokenprovider

Reply:
Yes absolutely, I had to choose whether to create` mtlsClient interface`(+ implementation) or use `creatorFunc` + standard http.Client implementation. For the sake of simplicity I chose creatorFunc. Do you see any drawbacks besides that there will always be just one default creator?